app/scripts/controllers/permissions/index.js

```diff
@@ -1 +1,561 @@
+      walletPrefix: WALLET_PREFIX,
+      restrictedMethods: (
+        Object.keys(this.externalRestrictedMethods)
+          .concat(Object.keys(this.pluginRestrictedMethods))
+      ),
+      store: this.store,
+      logStoreKey: LOG_STORE_KEY,
+      historyStoreKey: HISTORY_STORE_KEY,
+    }))
+    engine.push(this.permissions.providerMiddlewareFunction.bind(
+      this.permissions, { origin }
+    ))
+    return asMiddleware(engine)
+  }
+
+  /**
+   * Create middleware for prevent non-plugins from accessing methods only available to plugins
+   */
+  createPluginMethodRestrictionMiddleware (isPlugin) {
+    return createAsyncMiddleware(async (req, res, next) => {
+      if (typeof req.method !== 'string') {
+        res.error = ethErrors.rpc.invalidRequest({ data: req })
+        return // TODO:json-rpc-engine
+      }
+
+      if (pluginRestrictedMethodDescriptions[req.method] && !isPlugin) {
+        res.error = ethErrors.rpc.methodNotFound({ data: req.method })
+        return
+      }
+
+      return next()
+    })
+  }
+
+  /**
+   * @param {string} origin - The external domain id.
+   * @param {Object} requestedPlugins - The names of the requested plugin permissions.
+   */
+  async installPlugins (origin, requestedPlugins) {
+
+    const existingPerms = this.permissions.getPermissionsForDomain(origin).reduce(
+      (acc, p) => {
+        acc[p.parentCapability] = true
+        return acc
+      }, {}
+    )
+
+    const result = {}
+
+    // use a for-loop so that we can return an object and await the resolution
+    // of each call to processRequestedPlugin
+    for (const pluginName of Object.keys(requestedPlugins)) {
+
+      const permissionName = PLUGIN_PREFIX + pluginName
+      result[pluginName] = {
+        permission: permissionName,
+      }
+
+      // only allow the installation of permitted plugins
+      if (!existingPerms[permissionName]) {
+
+        result[pluginName].error = ethErrors.provider.unauthorized(
+          `Not authorized to install plugin '${pluginName}'. Please request the permission for the plugin before attempting to install it.`
+        )
+      } else {
+
+        // attempt to install and run the plugin, storing any errors that
+        // occur during the process
+        result[pluginName] = {
+          ...result[pluginName],
+          ...(await this.pluginsController.processRequestedPlugin(pluginName)),
+        }
+      }
+    }
+    return result
+  }
+
+  /**
+   * Returns the accounts that should be exposed for the given origin domain,
+   * if any. This method exists for when a trusted context needs to know
+   * which accounts are exposed to a given domain.
+   *
+   * Do not use in untrusted contexts; just send an RPC request.
+   *
+   * @param {string} origin
+   */
+  getAccounts (origin) {
+    return new Promise((resolve, _) => {
+      const req = { method: 'eth_accounts' }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
```

Comment:
Ah, you mean storing the `PermissionsController` middleware engines weakly by origin? I'm definitely not against it; that's what we do in `LoginPerSite` in `metamask-controller` in order to send notifications to external domains.

However, I think what we're currently doing there is more performant, because it bypasses the rest of the middleware stack to just hit the `CapabilitiesController`'s middleware function. Or am I missing something?